### PR TITLE
Introduce non-shipped System.Runtime.Internal assembly

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Introduce non-shipped System.Runtime.Internal assembly for internal test usage only and remove <ReferenceSystemPrivateCoreLib> support from our test system. Any internal APIs (primarily in the hosting space) that are needed for testing that we don't want to expose publicly should be exposed via System.Runtime.Internal for our testing infrastructure to reference.

This allows us to simplify our testing infrastructure to not have to handle tests that reference only CoreLib and to not have to build CoreLib in our test build job.

### CODE REVIEW

File: src/tests/Interop/COM/Activator/Program.cs [hunk 1]
```diff
@@ -12,7 +12,19 @@ namespace Activator
     using TestLibrary;
     using Xunit;
 
-    using Console = Internal.Console;
+    [ComImport]
+    [ComVisible(false)]
+    [Guid("00000001-0000-0000-C000-000000000046")]
+    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    public interface IClassFactory
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/Activator/Program.cs:19]:
```suggestion
    interface IClassFactory
```

**Review by AaronRobinsonMSFT — COMMENTED** [review]
This looks good. I'd like to see some indication for how we denote the project is only for testing and never designed to be shipped. Thoughts?

**Review by trylek — APPROVED** [review]
Awesome cleanup, thanks so much Jeremy!

**Comment by trylek** [comment]
@AaronRobinsonMSFT - I think that "never designed to be shipped" is impossible to define. The newly introduced S.R.I just adds support for stuff that tests need but the product doesn't have a reason or desire to ship [yet, potentially].

**Comment by jkoritzinsky** [comment]
We already have validation that we don't ship files that we don't expect in the Shared Framework SDK, so we don't need to worry about it implicitly being included there. Out-of-box packaging is an explicit opt-in, but just in case I've added an explicit MSBuild error if someone tries to pack the assembly into a NuGet package on its own.

**Comment by AaronRobinsonMSFT** [comment]
>We already have validation that we don't ship files that we don't expect in the Shared Framework SDK, so we don't need to worry about it implicitly being included there. 

This works for me. I much prefer having that.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
>  I think that "never designed to be shipped" is impossible to define.

I suppose. The point though is we need "expected to ship" or "not expected to ship" or at least the concept. In this case a verification step for the SDK handles the "expected to ship" verification so that means we intentionally say "ship it" somewhere and that was my general concern. I'm happy as long as we have a validation in one direction.

**Comment by jkotas** [comment]
My personal preference would be to test the internal interfaces using reflection instead of introducing the internal contracts.  IIRC, I had same feedback when some of these were introduced. It also matches how we are generally testing the internal details under libraries - we not build internal contracts for those.

**Comment by AaronRobinsonMSFT** [comment]
> My personal preference would be to test the internal interfaces using reflection instead of introducing the internal contracts. 

@jkotas I recall that. I think the biggest annoyance there is fiddling with the Reflection API. In this case, we would need to create a new instance of types and set fields. I guess this could be performed with `dynamic`. I will say that simple static function exports should be much easier with C# function pointers now. Can you refresh us on the reasoning again? Is there a cost we aren't considering?

**Comment by jkotas** [comment]
> I think the biggest annoyance there is fiddling with the Reflection API.

If you count number of lines, I think it would be less than the annoyance of introducing the internal contracts. `dynamic` can help, but it should not be too bad even without dynamic (create a few helper methods).

ICastable is annoying (you would have to use a bit of reflection.emit for it), but I hope that it will go away once we get the new source generator for COM.

> Can you refresh us on the reasoning again?

Consistency. If one wants to test internal details - what is going to be our guideline to decide when to use reflection vs. introducing internal contracts? We use reflection everywhere, except a few places.

**Comment by jkoritzinsky** [comment]
There's also the (temporary) issue of testing ICastable, which we don't expose publicly. The only option for implementing that specially would be through IL Emit, which I don't think many of us would want to maintain or own.

**Comment by jkoritzinsky** [comment]
For me, the reason that these contracts are different than our usual internal contracts is that they are `public` from System.Private.CoreLib, whereas all other contracts that we test are `internal`. As a result, we don't need to use reflection or InternalsVisibleTo to be able to access them easily. One of the main reasons we moved to using reflection in the libraries layer for testing internal details is because we wanted to remove all InternalsVisibleTo usage in the repo (or at least as much as possible) as it is very linker-unfriendly. Since these APIs are already `public`, they don't need InternalsVisibleTo or reflection to be usable, so exposing them through a ref assembly is more reasonable (at least to me).

**Comment by AaronRobinsonMSFT** [comment]
> they are public from System.Private.CoreLib

I think that is merely because using Reflection was deemed annoying at the time. It can't be used as the singular justification because marking them `public` is only for testing. They can be marked `internal` and the product should "just work" I believe.

**Comment by jkotas** [comment]
> For me, the reason that these contracts are different than our usual internal contracts is that they are public from System.Private.CoreLib, whereas all other contracts that we test are internal

I am not sure what are the other internal contracts you have in mind. For example, we have number of internal methods that support debugger, they are tested via reflection, and that form the contract between the runtime and the debugger.

If you flip these methods touched in this PR to internal, I believe that the product is still going to work fine, just the tests are going to break. (ICastable is an exception that will go away.)

**Comment by jkoritzinsky** [comment]
Does the hosting API surface (coreclr_get_delegate) support getting `internal` APIs? I didn't think it supported that, I thought they had to be `public`. If they can be `internal`, then I'm much more amenable to switching to using reflection for testing.

**Comment by jkotas** [comment]
> Does the hosting API surface (coreclr_get_delegate) support getting internal APIs? I

I think so. I have just checked - I do not see any visibility checks.

**Comment by jkoritzinsky** [comment]
I'll go down the route of converting our tests to use reflection.

For ICastable, I've got an idea to avoid ref-emit and I'll see how people feel about it. Since ICastable will go likely go away in the .NET 7 timeframe anyway, we have some more flexibility in what we do in the interim.

**Comment by jkoritzinsky** [comment]
Closing in favor of #61802

**Pull request closed without merging**